subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/worker/DeviceWorker.kt

```diff
@@ -55,52 +55,55 @@ internal class DeviceWorker(
+        try {
 
-            deviceListener.onIntentionReceived(device, intention)
+            for (intention in intentionsRouter.observeIntentions(state)) {
 
-            when (val status = device.deviceStatus()) {
+                deviceListener.onIntentionReceived(device, intention)
 
-                is Freeze -> {
-                    onDeviceDieOnIntention(intention, status.reason)
-                    return@launch
-                }
+                when (val status = device.deviceStatus()) {
 
-                is Alive -> {
+                    is Freeze -> {
+                        onDeviceDieOnIntention(intention, status.reason)
+                        return@launch
+                    }
 
-                    val (newState, preparingError) = prepareDeviceState(
-                        currentState = state,
-                        intendedState = intention.state
-                    )
+                    is Alive -> {
 
-                    when {
+                        val (newState, preparingError) = prepareDeviceState(
+                            currentState = state,
+                            intendedState = intention.state
+                        )
 
-                        preparingError != null -> {
-                            onDeviceDieOnIntention(intention, preparingError)
-                            return@launch
-                        }
+                        when {
+
+                            preparingError != null -> {
+                                onDeviceDieOnIntention(intention, preparingError)
+                                return@launch
+                            }
 
-                        newState != null -> {
-                            state = newState
+                            newState != null -> {
+                                state = newState
 
-                            deviceListener.onStatePrepared(device, newState)
+                                deviceListener.onStatePrepared(device, newState)
 
-                            deviceListener.onTestStarted(device, intention)
+                                deviceListener.onTestStarted(device, intention)
 
-                            val result = executeAction(action = intention.action)
+                                val result = executeAction(action = intention.action)
 
-                            deviceListener.onTestCompleted(
-                                device = device,
-                                intention = intention,
-                                result = result
-                            )
+                                deviceListener.onTestCompleted(
+                                    device = device,
+                                    intention = intention,
+                                    result = result
+                                )
+                            }
                         }
                     }
                 }
             }
+        } finally {
```

Comment:
Sometimes `onFinished` was not called because of `Channel`s cancelling. But it's important to collect utilization statistics

Reply:
What about tests for this case?

Reply:
Added a test for this